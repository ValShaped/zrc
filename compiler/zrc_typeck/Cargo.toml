[package]
name = "zrc_typeck"
version = "0.1.0"
edition = "2024"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
derive_more = { version = "2.0.1", features = ["display"] }
indexmap = "2.2.6"
zrc_diagnostics = { version = "0.1.0", path = "../zrc_diagnostics" }
zrc_parser = { version = "0.1.0", path = "../zrc_parser" }
zrc_utils = { version = "0.1.0", path = "../zrc_utils" }
