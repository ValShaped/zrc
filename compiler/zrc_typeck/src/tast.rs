//! Type-checked AST for the Zirco programming language
//!
//! This is similar to the [standard AST](zrc_parser::ast), but with type
//! information attached to it.

pub mod expr;
pub mod stmt;
pub mod ty;
