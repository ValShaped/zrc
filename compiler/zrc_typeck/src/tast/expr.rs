//! Expression representation for the Zirco [TAST](super)

use std::fmt::Display;

/// An [expression kind](TypedExprKind) with its yielded [result type](super::ty::Type) attached to it.
#[derive(PartialEq, Debug, Clone)]
#[allow(clippy::module_name_repetitions)]
pub struct TypedExpr(pub super::ty::Type, pub TypedExprKind);

/// The kind of a [`TypedExpr`]
#[derive(PartialEq, Debug, Clone)]
pub enum TypedExprKind {
    /// `a, b`
    Comma(Box<TypedExpr>, Box<TypedExpr>),

    /// `a = b`
    Assignment(Box<TypedExpr>, Box<TypedExpr>),

    /// `!x`
    UnaryNot(Box<TypedExpr>),
    /// `~x`
    UnaryBitwiseNot(Box<TypedExpr>),
    /// `-x`
    UnaryMinus(Box<TypedExpr>),
    /// `&x`
    UnaryAddressOf(Box<TypedExpr>),
    /// `*x`
    UnaryDereference(Box<TypedExpr>),

    /// `a[b]`
    Index(Box<TypedExpr>, Box<TypedExpr>),
    /// `a.b`
    Dot(Box<TypedExpr>, String),
    /// `a->b`
    Arrow(Box<TypedExpr>, String),
    /// `a(b, c, d, ...)`
    Call(Box<TypedExpr>, Vec<TypedExpr>),

    /// `a ? b : c`
    Ternary(Box<TypedExpr>, Box<TypedExpr>, Box<TypedExpr>),

    /// `a && b`
    LogicalAnd(Box<TypedExpr>, Box<TypedExpr>),
    /// `a || b`
    LogicalOr(Box<TypedExpr>, Box<TypedExpr>),

    /// `a == b`
    Equals(Box<TypedExpr>, Box<TypedExpr>),
    /// `a != b`
    NotEquals(Box<TypedExpr>, Box<TypedExpr>),

    /// `a & b`
    BitwiseAnd(Box<TypedExpr>, Box<TypedExpr>),
    /// `a | b`
    BitwiseOr(Box<TypedExpr>, Box<TypedExpr>),
    /// `a ^ b`
    BitwiseXor(Box<TypedExpr>, Box<TypedExpr>),

    /// `a > b`
    GreaterThan(Box<TypedExpr>, Box<TypedExpr>),
    /// `a >= b`
    GreaterThanOrEqualTo(Box<TypedExpr>, Box<TypedExpr>),
    /// `a < b`
    LessThan(Box<TypedExpr>, Box<TypedExpr>),
    /// `a <= b`
    LessThanOrEqualTo(Box<TypedExpr>, Box<TypedExpr>),

    /// `a >> b`
    BitwiseRightShift(Box<TypedExpr>, Box<TypedExpr>),
    /// `a << b`
    BitwiseLeftShift(Box<TypedExpr>, Box<TypedExpr>),

    /// `a + b`
    Addition(Box<TypedExpr>, Box<TypedExpr>),
    /// `a - b`
    Subtraction(Box<TypedExpr>, Box<TypedExpr>),

    /// `a * b`
    Multiplication(Box<TypedExpr>, Box<TypedExpr>),
    /// `a / b`
    Division(Box<TypedExpr>, Box<TypedExpr>),
    /// `a % b`
    Modulo(Box<TypedExpr>, Box<TypedExpr>),

    /// `x as T`
    Cast(Box<TypedExpr>, super::ty::Type),

    /// Any numeric literal.
    NumberLiteral(String),
    /// Any string literal.
    StringLiteral(String),
    /// Any identifier.
    Identifier(String),
    /// Any boolean literal.
    BooleanLiteral(bool),
}

impl Display for TypedExprKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::NumberLiteral(n) => write!(f, "{n}"),
            Self::StringLiteral(s) => write!(f, "{s}"),
            Self::Identifier(i) => write!(f, "{i}"),
            Self::BooleanLiteral(b) => write!(f, "{b}"),
            Self::Assignment(l, r) => write!(f, "{l} = {r}"),
            Self::Addition(l, r) => write!(f, "{l} + {r}"),
            Self::Subtraction(l, r) => write!(f, "{l} - {r}"),
            Self::Multiplication(l, r) => write!(f, "{l} * {r}"),
            Self::Division(l, r) => write!(f, "{l} / {r}"),
            Self::Modulo(l, r) => write!(f, "{l} % {r}"),
            Self::BitwiseAnd(l, r) => write!(f, "{l} & {r}"),
            Self::BitwiseOr(l, r) => write!(f, "{l} | {r}"),
            Self::BitwiseXor(l, r) => write!(f, "{l} ^ {r}"),
            Self::BitwiseLeftShift(l, r) => write!(f, "{l} << {r}"),
            Self::BitwiseRightShift(l, r) => write!(f, "{l} >> {r}"),
            Self::GreaterThan(l, r) => write!(f, "{l} > {r}"),
            Self::GreaterThanOrEqualTo(l, r) => write!(f, "{l} >= {r}"),
            Self::LessThan(l, r) => write!(f, "{l} < {r}"),
            Self::LessThanOrEqualTo(l, r) => write!(f, "{l} <= {r}"),
            Self::Equals(l, r) => write!(f, "{l} == {r}"),
            Self::NotEquals(l, r) => write!(f, "{l} != {r}"),
            Self::LogicalAnd(l, r) => write!(f, "{l} && {r}"),
            Self::LogicalOr(l, r) => write!(f, "{l} || {r}"),
            Self::Comma(l, r) => write!(f, "{l}, {r}"),
            Self::UnaryNot(e) => write!(f, "!{e}"),
            Self::Cast(x, t) => write!(f, "{x} as {t}"),
            Self::UnaryBitwiseNot(e) => write!(f, "~{e}"),
            Self::UnaryMinus(e) => write!(f, "-{e}"),
            Self::UnaryAddressOf(e) => write!(f, "&{e}"),
            Self::UnaryDereference(e) => write!(f, "*{e}"),
            Self::Arrow(l, r) => write!(f, "{l}->{r}"),
            Self::Ternary(a, b, c) => write!(f, "{a} ? {b} : {c}"),
            Self::Index(a, b) => write!(f, "{a}[{b}]"),
            Self::Dot(a, b) => write!(f, "{a}.{b}"),
            Self::Call(a, b) => write!(
                f,
                "{a}({})",
                b.iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
        }
    }
}

impl Display for TypedExpr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "(({}) as ({}))", self.1, self.0)
    }
}
