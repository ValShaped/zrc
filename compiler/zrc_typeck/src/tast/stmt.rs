//! Statement representation for the Zirco [TAST](super)

use std::fmt::Display;

use indexmap::IndexMap;

use super::{expr::TypedExpr, ty::Type};

/// A declaration created with `let`.
#[derive(Debug, Clone, PartialEq)]
pub struct LetDeclaration {
    /// The name of the identifier.
    pub name: String,
    /// The type of the new symbol. If set to [`None`], the type will be
    /// inferred.
    pub ty: Type, // types are definite after inference
    /// The value to associate with the new symbol.
    pub value: Option<TypedExpr>,
}

impl Display for LetDeclaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match &self.value {
            Some(v) => write!(f, "{}: {} = {}", self.name, self.ty, v),
            None => write!(f, "{}: {}", self.name, self.ty),
        }
    }
}

/// The enum representing all of the different kinds of statements in Zirco
/// after type checking
#[derive(Debug, Clone, PartialEq)]
#[allow(clippy::module_name_repetitions)]
pub enum TypedStmt {
    // all of the Box<Stmt>s for "possibly blocks" have been desugared into vec[single stmt] here
    // (basically if (x) y has become if (x) {y})
    /// `if (x) y` or `if (x) y else z`
    IfStmt(TypedExpr, Vec<TypedStmt>, Option<Vec<TypedStmt>>),
    /// `while (x) y`
    WhileStmt(TypedExpr, Vec<TypedStmt>),
    /// `for (init; cond; post) body`
    ForStmt {
        /// Runs once before the loop starts.
        init: Option<Box<Vec<LetDeclaration>>>,
        /// Runs before each iteration of the loop. If this evaluates to
        /// `false`, the loop will end. If this is [`None`], the loop
        /// will run forever.
        cond: Option<TypedExpr>,
        /// Runs after each iteration of the loop.
        post: Option<TypedExpr>,
        /// The body of the loop.
        body: Vec<TypedStmt>,
    },
    /// `{ ... }`
    BlockStmt(Vec<TypedStmt>),
    /// `x;`
    ExprStmt(TypedExpr),
    /// `;`
    EmptyStmt,
    /// `continue;`
    ContinueStmt,
    /// `break;`
    BreakStmt,
    /// `return;` or `return x;`
    ReturnStmt(Option<TypedExpr>),
    /// A let declaration
    DeclarationList(Vec<LetDeclaration>),
}

/// A struct or function declaration at the top level of a file
#[derive(Debug, Clone, PartialEq)]
pub enum TypedDeclaration {
    /// A declaration of a function
    FunctionDeclaration {
        /// The name of the function.
        name: String,
        /// The parameters of the function.
        parameters: Vec<ArgumentDeclaration>,
        /// The return type of the function. If set to [`None`], the function is
        /// void.
        return_type: Option<Type>,
        /// The body of the function. If set to [`None`], this is an extern
        /// declaration.
        body: Option<Vec<TypedStmt>>,
    },
    /// A named declaration for a `struct`.
    StructDeclaration {
        /// The name of the newtype.
        name: String,
        /// The key-value pairs of the struct. Ordered by declaration order.
        fields: IndexMap<String, super::ty::Type>,
    },
}

impl Display for TypedDeclaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: Some(r),
                body: Some(body),
            } => write!(
                f,
                "fn {name}({}) -> {r} {{\n{}\n}}",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", "),
                body.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<String>>()
                    .join("\n")
            ),
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: Some(r),
                body: None,
            } => write!(
                f,
                "fn {name}({}) -> {r};",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: None,
                body: Some(body),
            } => write!(
                f,
                "fn {name}({}) {{\n{}\n}}",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", "),
                body.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<String>>()
                    .join("\n")
            ),
            Self::FunctionDeclaration {
                name,
                parameters,
                return_type: None,
                body: None,
            } => write!(
                f,
                "fn {name}({});",
                parameters
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<String>>()
                    .join(", ")
            ),
            Self::StructDeclaration { name, fields } => write!(
                f,
                "struct {name} {{\n{}\n}}",
                fields
                    .iter()
                    .map(|(name, ty)| format!("    {name}: {ty}"))
                    .collect::<Vec<_>>()
                    .join(",\n")
            ),
        }
    }
}

impl Display for TypedStmt {
    #[allow(clippy::too_many_lines)]
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::IfStmt(e, s1, Some(s2)) => write!(
                f,
                "if ({e}) {{\n{}\n}} else {{\n{}\n}}",
                s1.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n"),
                s2.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::IfStmt(e, s1, None) => write!(
                f,
                "if ({e}) {{\n{}\n}}",
                s1.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::WhileStmt(e, s) => write!(
                f,
                "while ({e}) {{\n{}\n}}",
                s.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::ForStmt {
                init,
                cond,
                post,
                body,
            } => write!(
                f,
                "for ({} {}; {}) {{\n{}\n}}",
                init.clone().map_or(String::new(), |x| format!(
                    "let {};",
                    x.iter()
                        .map(std::string::ToString::to_string)
                        .collect::<Vec<_>>()
                        .join(", ")
                )),
                cond.clone().map_or(String::new(), |x| x.to_string()),
                post.clone().map_or(String::new(), |x| x.to_string()),
                body.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::BlockStmt(s) => write!(
                f,
                "{{\n{}\n}}",
                s.iter()
                    .map(|stmt| stmt
                        .to_string()
                        .split('\n')
                        .map(|x| format!("    {x}"))
                        .collect::<Vec<_>>()
                        .join("\n"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            Self::ExprStmt(e) => write!(f, "{e};"),
            Self::EmptyStmt => write!(f, ";"),
            Self::ContinueStmt => write!(f, "continue;"),
            Self::BreakStmt => write!(f, "break;"),
            Self::ReturnStmt(Some(e)) => write!(f, "return {e};",),
            Self::ReturnStmt(None) => write!(f, "return;"),
            Self::DeclarationList(d) => write!(
                f,
                "let {};",
                d.iter()
                    .map(std::string::ToString::to_string)
                    .collect::<Vec<_>>()
                    .join(", ")
            ),
        }
    }
}

/// A special form of [`LetDeclaration`] used for function parameters.
#[derive(PartialEq, Debug, Clone)]
pub struct ArgumentDeclaration {
    /// The name of the parameter.
    pub name: String,
    /// The type of the parameter.
    pub ty: Type,
}
impl Display for ArgumentDeclaration {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.name, self.ty)
    }
}
