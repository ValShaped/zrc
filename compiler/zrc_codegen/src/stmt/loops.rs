//! Code generation for switch statements

use inkwell::{basic_block::BasicBlock, debug_info::DILexicalBlock};
use zrc_typeck::tast::{
    expr::TypedExpr,
    stmt::{LetDeclaration, TypedStmt},
};
use zrc_utils::span::Spanned;

use crate::{
    bb::BasicBlockAnd,
    ctx::{BlockCtx, FunctionCtx},
    expr::cg_expr,
    scope::CgScope,
    stmt::{LoopBreakaway, cg_block},
    unpack,
};

/// Code generates a for statement
#[allow(clippy::too_many_arguments, clippy::ref_option, clippy::box_collection)]
pub fn cg_for_stmt<'ctx, 'input, 'a>(
    cg: FunctionCtx<'ctx, 'a>,
    bb: BasicBlock<'ctx>,
    scope: &'a CgScope<'input, 'ctx>,
    lexical_block: DILexicalBlock<'ctx>,
    init: Option<Box<Vec<Spanned<LetDeclaration<'input>>>>>,
    cond: Option<TypedExpr<'input>>,
    post: Option<TypedExpr<'input>>,
    body: Spanned<Vec<TypedStmt<'input>>>,
) -> BasicBlock<'ctx> {
    // For loops generate a somewhat more complicated CFG, with a few parts.
    // The preheader, where `init` runs. Breaks to the header.
    // The header, where `cond` is checked and breaks to either the exit or the
    // body. The body, where most of the body runs. Breaks to
    // the latch. `break` transfers to the exit by force and `continue` transfers to
    // the latch by force. The latch, where `post` runs and
    // breaks back to the header The exit, which is the basic
    // block we return.

    // loops lie in an implicit subscope
    let mut scope = scope.clone();

    // The block we are currently in will become the preheader. Generate the `init`
    // code if there is any.
    if let Some(init) = init {
        super::let_decl::cg_let_declaration(cg, bb, &mut scope, lexical_block, *init);
    }

    let expr_cg = BlockCtx::new(cg, &scope, lexical_block);

    let header = cg.ctx.append_basic_block(cg.fn_value, "header");
    let body_bb = cg.ctx.append_basic_block(cg.fn_value, "body");
    let latch = cg.ctx.append_basic_block(cg.fn_value, "latch");
    let exit = cg.ctx.append_basic_block(cg.fn_value, "exit");

    // Branch to the header from the preheader.
    cg.builder
        .build_unconditional_branch(header)
        .expect("branch should generate successfully");

    // Generate the header.
    cg.builder.position_at_end(header);
    let header = cond.map_or_else(
        || {
            // If there is no condition, we always branch to the body.
            cg.builder
                .build_unconditional_branch(body_bb)
                .expect("branch should generate successfully");

            header
        },
        |cond| {
            let mut header = header;

            let cond = unpack!(header = cg_expr(expr_cg, header, cond));

            cg.builder
                .build_conditional_branch(cond.into_int_value(), body_bb, exit)
                .expect("branch should generate successfully");

            header
        },
    );

    // Generate the body.
    cg.builder.position_at_end(body_bb);
    let body_bb = cg_block(
        cg,
        body_bb,
        &scope,
        lexical_block,
        body,
        &Some(LoopBreakaway {
            on_break: exit,
            on_continue: latch,
        }),
    );

    // The body breaks to latch
    if body_bb.is_some() {
        cg.builder
            .build_unconditional_branch(latch)
            .expect("branch should generate successfully");
    }

    // Latch runs post and then breaks right back to the header.
    cg.builder.position_at_end(latch);
    if let Some(post) = post {
        cg_expr(expr_cg, latch, post);
    }

    cg.builder
        .build_unconditional_branch(header)
        .expect("branch should generate successfully");

    cg.builder.position_at_end(exit);

    exit
}

/// Code generates a while statement
#[allow(clippy::too_many_arguments, clippy::ref_option)]
pub fn cg_while_stmt<'ctx, 'input, 'a>(
    cg: FunctionCtx<'ctx, 'a>,
    scope: &'a CgScope<'input, 'ctx>,
    lexical_block: DILexicalBlock<'ctx>,
    cond: TypedExpr<'input>,
    body: Spanned<Vec<TypedStmt<'input>>>,
) -> BasicBlock<'ctx> {
    let expr_cg = BlockCtx::new(cg, scope, lexical_block);

    // While loops are similar to for loops but much simpler.
    // The preheader simply just breaks to the header.
    // The header checks the condition and breaks to the exit or the body.
    // The body simply breaks to the header.
    // The exit is the continued code

    // `break` => exit
    // `continue` => header

    let mut header = cg.ctx.append_basic_block(cg.fn_value, "header");

    let body_bb = cg.ctx.append_basic_block(cg.fn_value, "body");

    let exit = cg.ctx.append_basic_block(cg.fn_value, "exit");

    cg.builder
        .build_unconditional_branch(header)
        .expect("branch should generate successfully");

    cg.builder.position_at_end(header);

    let cond = unpack!(header = cg_expr(expr_cg, header, cond));

    cg.builder
        .build_conditional_branch(cond.into_int_value(), body_bb, exit)
        .expect("branch should generate successfully");

    cg.builder.position_at_end(body_bb);

    let body_bb = cg_block(
        cg,
        body_bb,
        scope,
        lexical_block,
        body,
        &Some(LoopBreakaway {
            on_break: exit,
            on_continue: header,
        }),
    );

    if body_bb.is_some() {
        cg.builder
            .build_unconditional_branch(header)
            .expect("branch should generate successfully");
    }

    cg.builder.position_at_end(exit);
    exit
}

/// Code generates a do..while statement
#[allow(clippy::too_many_arguments, clippy::ref_option)]
pub fn cg_do_while_stmt<'ctx, 'input, 'a>(
    cg: FunctionCtx<'ctx, 'a>,
    scope: &'a CgScope<'input, 'ctx>,
    lexical_block: DILexicalBlock<'ctx>,
    body: Spanned<Vec<TypedStmt<'input>>>,
    cond: TypedExpr<'input>,
) -> BasicBlock<'ctx> {
    let expr_cg = BlockCtx::new(cg, scope, lexical_block);

    // `do..while` loops are slightly different from `while` loops.
    // the preheader breaks directly to the *body* and forces it to run at
    // least once. the body can then later break to the header which checks the
    // condition and will loop or exit.

    // `break` => exit
    // `continue` => header

    let body_bb = cg.ctx.append_basic_block(cg.fn_value, "body");
    let body_start = body_bb;

    let header = cg.ctx.append_basic_block(cg.fn_value, "header");

    let exit = cg.ctx.append_basic_block(cg.fn_value, "exit");

    cg.builder
        .build_unconditional_branch(body_bb)
        .expect("branch should generate successfully");

    cg.builder.position_at_end(body_bb);

    let body_bb = cg_block(
        cg,
        body_bb,
        scope,
        lexical_block,
        body,
        &Some(LoopBreakaway {
            on_break: exit,
            on_continue: header,
        }),
    );

    if body_bb.is_some() {
        cg.builder
            .build_unconditional_branch(header)
            .expect("branch should generate successfully");
    }

    cg.builder.position_at_end(header);

    let BasicBlockAnd { value: cond, .. } = cg_expr(expr_cg, header, cond);

    cg.builder
        .build_conditional_branch(cond.into_int_value(), body_start, exit)
        .expect("branch should generate successfully");

    cg.builder.position_at_end(exit);

    exit
}

#[cfg(test)]
mod tests {
    // Please read the "Common patterns in tests" section of crate::test_utils for
    // more information on how code generator tests are structured.

    use indoc::indoc;

    use crate::cg_snapshot_test;

    #[test]
    fn unreachable_statement_generates_properly() {
        cg_snapshot_test!(indoc! {"
            fn test(cond: bool) {
                let x = 7;
                if (x == 6) unreachable;
                else return;
            }
        "});
    }

    #[test]
    fn while_loops_along_with_break_and_continue_generate_as_expected() {
        cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: the proper while loop structure is created
                        while (get_bool()) {
                            // TEST: break jumps to the `end` block
                            if (get_bool()) break;
                            else {
                                // TEST: continue jumps to the header block
                                if (get_bool()) continue;
                                // TEST: otherwise, we proceed
                                else {}
                            }

                            // TEST: the loop jumps back to the header block
                        }

                        // TEST: ...and code generation properly continues.
                        return;
                    }
                "});
    }

    #[test]
    fn for_loops_along_with_break_and_continue_generate_as_expected() {
        cg_snapshot_test!(indoc! {"
                    fn get_int() -> i32;

                    fn test() {
                        // TEST: the proper while loop structure is created
                        for (let i = 0; i < get_int(); i += 1) {
                            // TEST: break jumps to the `end` block
                            if (i > get_int()) break;
                            else {
                                // TEST: continue jumps to the latch block
                                if (i < get_int()) continue;
                                else {}
                            }

                            // TEST: the loop jumps to the latch block which jumps back to the
                            // header
                        }
                    }
                "});
    }

    #[test]
    fn do_while_loops_generate_as_expected() {
        cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;

                    fn test() {
                        // TEST: the proper `do..while` loop structure is created
                        do {
                            get_bool(); // for fake side effects
                        } while (get_bool());
                    }
                "});
    }

    #[test]
    fn switch_statements_generate_as_expected() {
        cg_snapshot_test!(indoc! {"
                    fn get_bool() -> bool;
                    fn when_true();
                    fn when_false_a();
                    fn when_false_b();
                    fn when_default(x: i32);
                    fn post();


                    fn test() {
                        // TEST: the proper `switch` structure is created
                        switch (get_bool()) {
                            true => when_true();
                            false => {
                                when_false_a();
                                when_false_b();
                            }
                            default => {
                                let x = 2 + 2;
                                when_default(x);
                            }
                        }
                        post();
                    }
                "});
    }
}
