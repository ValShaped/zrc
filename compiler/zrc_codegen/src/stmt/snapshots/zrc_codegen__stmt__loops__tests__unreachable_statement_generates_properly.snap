---
source: compiler/zrc_codegen/src/stmt/loops.rs
description: "fn test(cond: bool) {\n    let x = 7;\n    if (x == 6) unreachable;\n    else return;\n}\n"
expression: resulting_ir
---
; ModuleID = 'test.zr'
source_filename = "test.zr"

define {} @test(i1 %0) !dbg !3 {
entry:
  %let_x = alloca i32, align 4, !dbg !9
  %arg_cond = alloca i1, align 1, !dbg !9
  store i1 %0, ptr %arg_cond, align 1, !dbg !9
  store i32 7, ptr %let_x, align 4, !dbg !11
  %load = load i32, ptr %let_x, align 4, !dbg !13
  %cmp = icmp eq i32 %load, 6, !dbg !14
  br i1 %cmp, label %then, label %then_else, !dbg !14

then:                                             ; preds = %entry
  unreachable, !dbg !15

then_else:                                        ; preds = %entry
  ret {} zeroinitializer, !dbg !17
}

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "zrc test runner", isOptimized: false, flags: "zrc --fake-args", runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false)
!2 = !DIFile(filename: "test.zr", directory: "/fake/path")
!3 = distinct !DISubprogram(name: "test", linkageName: "test", scope: null, file: !2, line: 1, type: !4, scopeLine: 1, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6, !8}
!6 = !DICompositeType(tag: DW_TAG_structure_type, name: "struct {}", scope: !2, file: !2, elements: !7)
!7 = !{}
!8 = !DIBasicType(name: "bool")
!9 = !DILocation(line: 1, column: 21, scope: !10)
!10 = distinct !DILexicalBlock(scope: !3, file: !2, line: 1, column: 21)
!11 = !DILocation(line: 2, column: 9, scope: !12)
!12 = distinct !DILexicalBlock(scope: !10, file: !2, line: 1, column: 21)
!13 = !DILocation(line: 3, column: 9, scope: !12)
!14 = !DILocation(line: 3, column: 14, scope: !12)
!15 = !DILocation(line: 3, column: 17, scope: !16)
!16 = distinct !DILexicalBlock(scope: !12, file: !2, line: 3, column: 17)
!17 = !DILocation(line: 4, column: 10, scope: !18)
!18 = distinct !DILexicalBlock(scope: !12, file: !2, line: 4, column: 10)
